### PR TITLE
Multi-Account Jazz Metrics API Changes

### PR DESCRIPTION
**Pull request by Johnbush**
### Requirements

While moving on to multi account feature, there is a need to get Metrics Data from non-primary Accounts. In order to fetch the metrics data we need the temporary access credentials of Non-Primary Account.

### Description of the Change

- Get Account Id from Service catalog
- Get temporary access creds using Assume role
- New instance of AWS object has been created with the temp creds to access the metrics data.

### CODE REVIEW

**Comment by SiniWilson** [comment]
travis is failing , please fix it

File: core/jazz_metrics/test/test.js [hunk 1]
```diff
@@ -581,6 +581,18 @@ describe('jazz_metrics', function () {
...
+      accessparams = {
+        "accessKeyId": "hewf73223e2h3u23y82332h23nd",
+        "secretAccessKey": "48584ji4rnfruihv849fjne",
+        "sessionToken": "nckjvrevr4343k4jioj43u94j3m4oi493f43jfm43fcn",
+      };
+      serviceData = {
+        "count": 1,
+        "services" :{
+          "region" : "us-east-1",
+          "iamRoleARN": "arn:aws:iam::102707241671:role/config21_basic_execution",
```
Comment (pspani) [inline core/jazz_metrics/test/test.js]: remove identifying information

Comment (SiniWilson) [reply core/jazz_metrics/test/test.js]: This is security violation, please squash the commit

File: core/jazz_metrics/components/utils.js [hunk 2]
```diff
@@ -240,26 +240,55 @@ function updateCloudfrontAsset(newAssetObj, relativeId) {
...
-  });
+function getCloudfrontCloudWatch(tempcreds , serviceMetaData) {
+  tempcreds.apiVersion = '2010-08-01';
+  tempcreds.region = serviceMetaData.services.region
+  var cloudwatch = new AWS.CloudWatch({tempcreds});
   return cloudwatch;
 }
 
+function AssumeRole(accountID) {
+  logger.info("Inside Assume Role")
```
Comment (SiniWilson) [inline core/jazz_metrics/components/utils.js]: Remove this log

File: core/jazz_metrics/components/utils.js [hunk 3]
```diff
@@ -240,26 +240,55 @@ function updateCloudfrontAsset(newAssetObj, relativeId) {
...
+  return new Promise((resolve, reject) => {
+    const sts = new AWS.STS({ region: process.env.REGION });
+    const params = {
+      RoleArn: 'arn:aws:iam::'+accountID+':role/'+global_config.STACK_PREFIX+'_platform_services',
```
Comment (SiniWilson) [inline core/jazz_metrics/components/utils.js]: RoleArn: `arn:aws:iam::${accountID}:role/${global_config.STACK_PREFIX}_platform_services`,

File: core/jazz_metrics/components/utils.js [hunk 4]
```diff
@@ -240,26 +240,55 @@ function updateCloudfrontAsset(newAssetObj, relativeId) {
...
+    };
+    sts.assumeRole(params, (err, data) => {
+      if (err) {
+        logger.error("Error Assuming Role with params:" + JSON.stringify(err));
+        reject({
+          "result": "serverError",
+          "message": "Unknown internal error occurred"
+        })
+      }else{
+        logger.info("The Temporary Acces details are:" + JSON.stringify(data));
```
Comment (SiniWilson) [inline core/jazz_metrics/components/utils.js]: logger.debug

File: core/jazz_metrics/config/dev-config.json [hunk 5]
```diff
@@ -1,6 +1,7 @@
 {
   "SERVICE_API_URL": "https://{conf-apikey}.execute-api.{conf-region}.amazonaws.com",
   "ASSETS_URL": "/dev/jazz/assets",
+  "SERVICE_URL": "/dev/jazz/services",
   "TOKEN_URL": "/dev/jazz/login",
```
Comment (SiniWilson) [inline core/jazz_metrics/config/dev-config.json]: "SERVICE_API_URL": "https://{conf-apikey}.execute-api.{conf-region}.amazonaws.com/dev",

And remove the dev from the rest of the values

File: core/jazz_metrics/config/prod-config.json [hunk 6]
```diff
@@ -1,6 +1,7 @@
 {
   "SERVICE_API_URL": "https://{conf-apikey}.execute-api.{conf-region}.amazonaws.com",
```
Comment (SiniWilson) [inline core/jazz_metrics/config/prod-config.json]: same as above comment

File: core/jazz_metrics/config/stg-config.json [hunk 7]
```diff
@@ -1,6 +1,7 @@
 {
   "SERVICE_API_URL": "https://{conf-apikey}.execute-api.{conf-region}.amazonaws.com",
   "ASSETS_URL": "/stg/jazz/assets",
+  "SERVICE_URL": "/stg/jazz/services",
```
Comment (SiniWilson) [inline core/jazz_metrics/config/stg-config.json]: same as above comment

File: core/jazz_metrics/index.js [hunk 8]
```diff
@@ -128,6 +141,37 @@ function getToken(config) {
     });
   });
 }
+function getserviceMetaData(config, eventBody, authToken){
+  return new Promise((resolve, reject) =>  {
+    var service_api_options = {
+      url: config.SERVICE_API_URL + config.SERVICE_URL + "?domain=" + eventBody.domain + "&service=" + eventBody.service + "&environment=" + eventBody.environment,
```
Comment (SiniWilson) [inline core/jazz_metrics/index.js]: Make use of ` key  and remove the +

File: core/jazz_metrics/index.js [hunk 9]
```diff
@@ -128,6 +141,37 @@ function getToken(config) {
...
+      async: true
+    };
+
+    request(service_api_options, (error, response, body) => {
+      if (error) {
+        reject(error);
+      } else {
+        if (response.statusCode && response.statusCode === 200) {
+          var responseBody = JSON.parse(body);
+          //var iamroleARN = responseBody.data.services.metadata.iamRoleARN
```
Comment (SiniWilson) [inline core/jazz_metrics/index.js]: Remove unwanted codes

File: core/jazz_metrics/index.js [hunk 10]
```diff
@@ -128,6 +141,37 @@ function getToken(config) {
...
+          //var AccountId = iamroleARN.split(':')[4];
+          resolve(responseBody.data)
+        }else{
+          logger.info("Service not found for this service, domain, environment. ", JSON.stringify(service_api_options));
```
Comment (SiniWilson) [inline core/jazz_metrics/index.js]: logger.debug

Comment (Johnbush) [reply core/jazz_metrics/index.js]: updated.

File: core/jazz_metrics/index.js (hunk 9 above)
Comment (Johnbush) [reply core/jazz_metrics/index.js]: updated

File: core/jazz_metrics/components/utils.js (hunk 2 above)
Comment (Johnbush) [reply core/jazz_metrics/components/utils.js]: updated.

File: core/jazz_metrics/components/utils.js (hunk 4 above)
Comment (Johnbush) [reply core/jazz_metrics/components/utils.js]: updated.

File: core/jazz_metrics/components/utils.js (hunk 3 above)
Comment (Johnbush) [reply core/jazz_metrics/components/utils.js]: This string has been removed on the newer commits.

File: core/jazz_metrics/config/stg-config.json (hunk 7 above)
Comment (Johnbush) [reply core/jazz_metrics/config/stg-config.json]: Updated.

File: core/jazz_metrics/index.js (hunk 8 above)
Comment (Johnbush) [reply core/jazz_metrics/index.js]: updated.

File: core/jazz_metrics/config/dev-config.json (hunk 5 above)
Comment (Johnbush) [reply core/jazz_metrics/config/dev-config.json]: Done also updated for stg and prod config too.

File: core/jazz_metrics/index.js [hunk 11]
```diff
@@ -46,16 +46,29 @@ function handler(event, context, cb) {
...
-        var finalObj = utils.massageData(res, eventBody);
-        return cb(null, responseObj(finalObj, eventBody));
+    var genValidation = exportable.genericValidation(event);
+    var token = exportable.getToken(config);
+    var valGenFields = validateUtils.validateGeneralFields(eventBody);
+    Promise.all([genValidation, token, valGenFields])
+    .then((results) => {
+      logger.info(results);
+      const authToken = results[1];
+      exportable.getserviceMetaData(config, eventBody, authToken)
```
[the remark this answers is not in the document]
Comment (pspani) [reply core/jazz_metrics/index.js]: as discussed, as the code is working and has been tested no changes required

**Pull request merged**